src/tiered/tiered_handle.c

```diff
@@ -8,63 +8,488 @@
+    }
+    WT_RET(__wt_buf_catfmt(session, tmp, ")"));
+
+    cfg[0] = WT_CONFIG_BASE(session, tiered_meta);
+    cfg[1] = orig_config;
+    cfg[2] = tmp->data;
+    WT_ERR(__wt_config_merge(session, cfg, NULL, &newconfig));
+    __wt_verbose(
+      session, WT_VERB_TIERED, "TIER_UPDATE_META: Update TIERED: %s %s", dhandle->name, newconfig);
+    WT_ERR(__wt_metadata_update(session, dhandle->name, newconfig));
+
+err:
+    __wt_free(session, newconfig);
+    __wt_scr_free(session, &tmp);
+    return (ret);
+}
+
+/*
+ * __tiered_switch --
+ *     Given a tiered table, make all the metadata updates underneath to switch to the next object.
+ *     The switch handles going from nothing to local-only, local-only to both local and shared, and
+ *     having shared-only and creating a local object. Must be single threaded.
+ */
+static int
+__tiered_switch(WT_SESSION_IMPL *session, const char *config)
+{
+    WT_DATA_HANDLE *dhandle;
+    WT_DECL_RET;
+    WT_TIERED *tiered;
+    bool need_local, need_object, need_tree, tracking;
+
+    dhandle = session->dhandle;
+    tiered = (WT_TIERED *)dhandle;
+    __wt_verbose(session, WT_VERB_TIERED, "TIER_SWITCH: called %s %s", dhandle->name, config);
+
+    /* We might only have a local file tree so far. */
+    WT_RET_NOTFOUND_OK(ret);
+
+    need_local = true;
+    need_object = tiered->tiers[WT_TIERED_INDEX_WRITEABLE] != NULL;
+    need_tree = need_object && tiered->tiers[WT_TIERED_INDEX_READONLY] == NULL;
+    /*
+     * There are four possibilities to our tiers configuration. In all of them we need to create
+     * a new local tier file object dhandle and add it as element index zero of the tiers array.
+     * Then we may or may not do other operations depending on the state otherwise. These are
+     * presented in order of increasing amount of work that needs to be done.
+     *   1. tiers=() - New and empty. We only need to add in the local file object.
+     *   2. tiers=("tier:...") - Existing shared tier only. Here too we only need to add
+     *      in the local file object.
+     *   3. tiers=("file:...", "tier:...") - Both local and shared tiers exist in the metadata.
+     *       We need to create and add the next local file object (N + 1) and create a shared
+     * object in the metadata for the current local file object (N).
+     *   4. tiers=("file:...") - Existing local tier only. We need to do all of the parts listed
+     * in the #3 above, and also create the shared tier metadata entry.
+     *
+     * The order we perform these operations is important. The order is:
+     *   1. Create the "object:" entry in metadata if needed.
+     *   2. Create the "tier:" entry in metadata if needed.
+     *   3. Create the new "file:" local entry in metadata.
+     *   4. Update the "tiered:" with new tiers and object number.
+     *   5. Meta tracking off to "commit" all the metadata operations.
```

Comment:
If we have meta tracking on, why is ordering important, at least for the first 4 steps?